File: MIDTERM/q2.cpp
// C++ program to find the missing integer 
// in N numbers when N bits are given 
#include <bits/stdc++.h> 
using namespace std; 
  
class BitInteger { 
private: 
    bool* bits; 
  
public: 
    static const int INTEGER_SIZE = 32; 
  
    BitInteger() 
    { 
        bits = new bool[INTEGER_SIZE]; 
    } 
  
    // Constructor to convert an integer 
    // variable into binary format 
    BitInteger(int value) 
    { 
        bits = new bool[INTEGER_SIZE]; 
  
        for (int j = 0; j < INTEGER_SIZE; j++) { 
  
            // The if statement will shift the 
            // original value j times. 
            // So that appropriate (INTEGER_SIZE - 1 -j)th 
            // bits will be either 0/1. 
            //  (INTEGER_SIZE - 1 -j)th bit for all 
            // j = 0 to INTEGER_SIZE-1 corresponds 
            // to  LSB to MSB respectively. 
            if (((value >> j) & 1) == 1) 
                bits[INTEGER_SIZE - 1 - j] = true; 
            else
                bits[INTEGER_SIZE - 1 - j] = false; 
        } 
    } 
    // Constructor to convert a 
    // string into binary format. 
    BitInteger(string str) 
    { 
        int len = str.length(); 
        int x = INTEGER_SIZE - len; 
        bits = new bool[INTEGER_SIZE]; 
  
        // If len = 4. Then x = 32 - 4 = 28. 
        // Hence iterate from 
        // bit 28 to bit 32 and just 
        // replicate the input string. 
        int i = 0; 
  
        for (int j = x; j <= INTEGER_SIZE && i < len; j++, i++) { 
            if (str[i] == '1') 
                bits[j] = true; 
            else
                bits[j] = false; 
        } 
    } 
  
    // this function fetches the kth bit 
    int fetch(int k) 
    { 
        if (bits[k]) 
            return 1; 
  
        return 0; 
    } 
  
    // this function will set a value 
    // of bit indicated by k to given bitValue 
    void set(int k, int bitValue) 
    { 
        if (bitValue == 0) 
            bits[k] = false; 
        else
            bits[k] = true; 
    } 
  
    // convert binary representation to integer 
    int toInt() 
    { 
        int n = 0; 
        for (int i = 0; i < INTEGER_SIZE; i++) { 
            n = n << 1; 
            if (bits[i]) 
                n = n | 1; 
        } 
        return n; 
    } 
}; 
  
// Function to find the missing number 
int findMissingFunc(list<BitInteger>& myList, int column) 
{ 
    // This means that we have processed 
    // the entire 32 bit binary number. 
    if (column < 0) 
        return 0; 
  
    list<BitInteger> oddIndices; 
    list<BitInteger> evenIndices; 
  
    for (BitInteger t : myList) { 
  
        // Initially column = LSB. So 
        // if LSB of the given number is 0, 
        // then the number is even and 
        // hence we add it to evenIndices list. 
        // else if LSB = 0 then add it to oddIndices list. 
        if (t.fetch(column) == 0) 
            evenIndices.push_back(t); 
        else
            oddIndices.push_back(t); 
    } 
  
    // Step 1 and Step 2 of the algorithm. 
    // Here we determine the LSB bit of missing number. 
  
    if (oddIndices.size() >= evenIndices.size()) 
  
        // LSB of the missing number is 0. 
        // Hence it is an even number. 
        // Step 3 and 4 of the algorithm 
        // (discarding all odd numbers) 
        return (findMissingFunc(evenIndices, column - 1)) << 1 | 0; 
  
    else
        // LSB of the missing number is 1. 
        // Hence it is an odd number. 
        // Step 3 and 4 of the algorithm 
        // (discarding all even numbers) 
        return (findMissingFunc(oddIndices, column - 1)) << 1 | 1; 
} 
  
// Function to return the missing integer 
int findMissing(list<BitInteger>& myList) 
{ 
    // Initial call is with given array and LSB. 
    return findMissingFunc(myList, BitInteger::INTEGER_SIZE - 1); 
} 
  
// Driver Code. 
int main() 
{ 
  
    // a corresponds to the input array which 
    // is a list of binary numbers 
    list<BitInteger> a = { BitInteger("0000"), BitInteger("0001"), 
                           BitInteger("0010"), BitInteger("0100"), 
                           BitInteger("0101") }; 
    int missing1 = findMissing(a); 
    cout << missing1 << "\n"; 
  
    return 0; 
} 

File: MIDTERM/q2.py
class myBinary:
    MAXBIT = 32
    __bits = list()

    def __init__(self, number):
        if type(number) == type(""):
            self.__bits = self.__strToBin(number)
        elif type(number) == type(1):
            self.__bits = self.__intToBin(number)

    # Returns the binary equivalent of the given string binary as bool list
    def __strToBin(self, number):
        tmp = list()
        for ele in range(self.MAXBIT-len(number)):
            tmp.append(False)
        for ele in number:
            tmp.append(ele=="1")
        return tmp

    # Returns the binary equivalent of the given integer number as bool list
    def __intToBin(self,number):
        return self.__strToBin(f'{number:b}')

    #string representation of the binary numbers
    def __str__(self):
        # print(self.__bits)
        res = ""
        for ele in self.__bits:
            res += str(int(ele))
        return res

    # returns the bit in the given index
    def carryBit(self, index):
        # print("carry:", index)
        if self.__bits[index]:
            return 1
        return 0


def findAbsentNumberCaller(binList, index):
    # print("index: ", index)
    if index<0:
        return 0
    
    oddIndices = list()
    evenIndices = list()

    for ele in binList:
        if ele.carryBit(index) == 0:
            evenIndices.append(ele)
        else:
            oddIndices.append(ele)
    if len(oddIndices) >= len(evenIndices):
        return findAbsentNumberCaller(evenIndices, index-1) << 1 | 0
    else:
        return findAbsentNumberCaller(oddIndices, index-1) << 1 | 1

def findAbsentNumber(binList, lenght):
    return findAbsentNumberCaller(binList, lenght-1)


binList = [myBinary(1), myBinary(4), myBinary(3),myBinary(5), 
            myBinary(0),myBinary(2),myBinary(7)]


# binList = [myBinary("0000"), myBinary("0001"),myBinary("0010"), 
#             myBinary("0100"),myBinary("0101")]

# for ele in binList:
#     print(ele)

print(findAbsentNumber(binList,myBinary.MAXBIT))